File: src/main/java/com/nekiplay/hypixelcry/utils/SpecialColor.java
package com.nekiplay.hypixelcry.utils;

import java.awt.*;

@SuppressWarnings("AvoidBritishSpelling")
public class SpecialColor {
    private static final int MIN_CHROMA_SECS = 1;
    private static final int MAX_CHROMA_SECS = 60;
    private static final SimpleTimeMark startTime = SimpleTimeMark.now();
    private static final Color DEFAULT_COLOR = new Color(0, 0, 0, 0); // Transparent black as default

    public static Color toSpecialColor(String str) {
        if (str == null || str.isEmpty()) {
            return DEFAULT_COLOR;
        }
        return new Color(toSpecialColorInt(str), true);
    }

    public static int toSpecialColorInt(String str) {
        if (str == null || str.isEmpty()) {
            return DEFAULT_COLOR.getRGB();
        }

        int[] components = decompose(str);
        int chroma = components[0];
        int alpha = components[1];
        int red = components[2];
        int green = components[3];
        int blue = components[4];

        float[] hsb = Color.RGBtoHSB(red, green, blue, null);
        float hue = hsb[0];
        float sat = hsb[1];
        float bri = hsb[2];

        float adjustedHue;
        if (chroma > 0) {
            float hueChange = (startTime.passedSince().toMillis() / 1000f / chromaSpeed(chroma) % 1);
            adjustedHue = hue + hueChange;
            if (adjustedHue < 0) {
                adjustedHue += 1f;
            }
        } else {
            adjustedHue = hue;
        }

        return (alpha & 0xFF) << 24 | (Color.HSBtoRGB(adjustedHue, sat, bri) & 0x00FFFFFF);
    }

    private static int[] decompose(String csv) {
        if (csv == null || csv.isEmpty()) {
            return new int[5]; // Returns array of zeros
        }

        String[] parts = csv.split(":");
        int[] result = new int[5];
        for (int i = 0; i < parts.length && i < 5; i++) {
            try {
                result[i] = Integer.parseInt(parts[i]);
            } catch (NumberFormatException e) {
                result[i] = 0;
            }
        }
        return result;
    }

    private static float chromaSpeed(int speed) {
        return (255 - speed) / 254f * (MAX_CHROMA_SECS - MIN_CHROMA_SECS) + MIN_CHROMA_SECS;
    }
}

File: src/main/kotlin/com/nekiplay/hypixelcry/utils/SpecialColor.kt
package com.nekiplay.hypixelcry.utils

import java.awt.Color

object SpecialColor {
    private const val MIN_CHROMA_SECS = 1
    private const val MAX_CHROMA_SECS = 60
    private val startTime = System.currentTimeMillis()

    @JvmStatic
    fun String.toSpecialColor() = Color(toSpecialColorInt(), true)

    @JvmStatic
    fun String.toSpecialColorInt(): Int {
        val components = decompose(this)
        val chroma = components.getOrElse(0) { 0 }
        val alpha = components.getOrElse(1) { 255 }
        val red = components.getOrElse(2) { 0 }
        val green = components.getOrElse(3) { 0 }
        val blue = components.getOrElse(4) { 0 }

        val (hue, sat, bri) = Color.RGBtoHSB(red, green, blue, null)

        val adjustedHue = if (chroma > 0) (hue + ((System.currentTimeMillis() - startTime) / 1000f / chromaSpeed(chroma) % 1)).let {
            if (it < 0) it + 1f else it
        } else hue

        return (alpha and 0xFF) shl 24 or (Color.HSBtoRGB(adjustedHue, sat, bri) and 0x00FFFFFF)
    }

    @JvmStatic
    fun String.toSpecialColorIntNoAlpha(): Int {
        val components = decompose(this)
        val chroma = components.getOrElse(0) { 0 }
        val red = components.getOrElse(if (components.size == 4) 1 else 2) { 0 }
        val green = components.getOrElse(if (components.size == 4) 2 else 3) { 0 }
        val blue = components.getOrElse(if (components.size == 4) 3 else 4) { 0 }

        val (hue, sat, bri) = Color.RGBtoHSB(red, green, blue, null)

        val adjustedHue = if (chroma > 0) (hue + ((System.currentTimeMillis() - startTime) / 1000f / chromaSpeed(chroma) % 1)).let {
            if (it < 0) it + 1f else it
        } else hue

        return Color.HSBtoRGB(adjustedHue, sat, bri)
    }

    @JvmStatic
    fun String.toSpecialColorFloatArray(): FloatArray {
        val colorInt = toSpecialColor()
        return floatArrayOf(
            colorInt.red / 255f,
            colorInt.green / 255f,
            colorInt.blue / 255f,
            colorInt.alpha / 255f
        )
    }

    private fun decompose(csv: String) = csv.split(":").mapNotNull { it.toIntOrNull() }.toIntArray()
    private fun chromaSpeed(speed: Int) = (255 - speed) / 254f * (MAX_CHROMA_SECS - MIN_CHROMA_SECS) + MIN_CHROMA_SECS
}